spring-cloud-dataflow-module-deployers/spring-cloud-dataflow-module-deployer-kubernetes/src/main/java/org/springframework/cloud/dataflow/module/deployer/kubernetes/KubernetesModuleDeployer.java

```diff
@@ -0,0 +1,202 @@
+
+		logger.debug("Deploying module: {}", createKubernetesName(id));
+
+		int externalPort = 8080;
+		// we also create a service in case have an '--port' or '--server.port' argument (source or sink)
+		Map<String, String> parameters = request.getDefinition().getParameters();
+		if (parameters.containsKey(PORT_KEY)) {
+			externalPort = Integer.valueOf(parameters.get(PORT_KEY));
+			createService(id, request, externalPort);
+		} else if (parameters.containsKey(SERVER_PORT_KEY)) {
+			externalPort = Integer.valueOf(parameters.get(SERVER_PORT_KEY));		
+			createService(id, request, externalPort);
+		}		
+		createReplicationController(id, request, externalPort);				
+		return id;	
+	}
+
+	@Override
+	public void undeploy(ModuleDeploymentId id) {
+		String name = createKubernetesName(id);
+		logger.debug("Undeploying module: {}", name);
+
+		try {
+			kubernetesClient.replicationControllers().withName(name).delete();
+			kubernetesClient.services().withName(name).delete();
+		} catch (KubernetesClientException e) {
+			logger.error(e.getMessage(), e);
+			throw new RuntimeException(e);
+		}
+	}
+
+	@Override
+	public ModuleStatus status(ModuleDeploymentId id) {
+		String name = createKubernetesName(id);
+		logger.debug("Querying module status: {}", name);
+
+		try {
+			// The only really interesting status is coming from the containers and pods.
+			// The service and the RC don't have "realtime" status info.
+			PodList list = kubernetesClient.pods().withLabels(createIdMap(id)).list();
+			return buildModuleStatus(id, list);
+		}
+		catch (KubernetesClientException e) {
+			logger.warn(e.getMessage(), e);
+			return buildModuleStatus(id, null);
+		}
+	}
+	
+	@Override
+	public Map<ModuleDeploymentId, ModuleStatus> status() {
+		Map<ModuleDeploymentId, ModuleStatus> result = new HashMap<>();
+		
+		ReplicationControllerList list = 
+				kubernetesClient.replicationControllers()
+				.withLabel(SPRING_MARKER_KEY, SPRING_MARKER_VALUE).list();
+		
+		for (ReplicationController rc : list.getItems()) {
+			Map<String, String> labels = rc.getMetadata().getLabels();			
+			String group = labels.get(SCSM_GROUP_KEY);
+			String label = labels.get(SCSM_LABEL_KEY);			
+			ModuleDeploymentId id = new ModuleDeploymentId(group, label);
+			
+			PodList pods = kubernetesClient.pods().withLabels(labels).list();			
+			result.put(id, buildModuleStatus(id, pods));	
+		}		
+		return result;		
+	}
+
+	private ReplicationController createReplicationController(ModuleDeploymentId id, ModuleDeploymentRequest request, int externalPort) {
+        ReplicationController rc = new ReplicationControllerBuilder()
+            .withNewMetadata()
+            	.withName(createKubernetesName(id))  // does not allow . in the name
+            	.withLabels(createIdMap(id))
+            	.addToLabels(SPRING_MARKER_KEY, SPRING_MARKER_VALUE)
+        		.addToLabels(SCSM_ARTIFACT_ID, request.getCoordinates().getArtifactId())
```

Comment:
I wanted label because I can query them (labels are supposed to be "identifying"). Annotations is something that I cannot really query. 

Labels have the extra advantage that I can also work with them with `kubectl`. If I'm the admin of a large cluster, they come in handy to query all to dataflows that use a module of a certain version and perform a rolling upgrade.

`kubectl get pods -l scsm-label=http scsm-version=1.1.0.SNAPSHOT`